### PR TITLE
Design refresh

### PR DESCRIPTION
**Pull request by randyaskin**
This PR provides a rebuild of the px-dropdown component using iron-dropdown and iron-selector. It simplifies the API by removing the px-dropdown-content subcomponent and providing all of the configuration properties on the px-dropdown component itself. It is a rather large departure from the previous API, and will have to be updated and tested in px-demo, px-sass-doc, px-data-table, px-inbox and any other components that currently use px-dropdown. Also needs to be updated with the new buttons and form input designs as part of the current design refresh activities and signed off on by the designers.

Addresses issues #34 and #40.

### CODE REVIEW

File: sass/px-dropdown.scss [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+}
+:host.validation-warning .dropdown-trigger {
+  border-color: var(--px-validation-warning-text-color, $status-orange3);
+  color: inherit;
+}
+:host.validation-success .dropdown-trigger {
+  border-color: var(--px-validation-success-text-color, $status-green3);
+  color: inherit;
+}
+.dropdown-trigger {
```
Comment (mdwragg) [inline sass/px-dropdown.scss]: Suggest we BEM the names up here `.dropdown__trigger`?

File: sass/px-dropdown.scss [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  text-align: left;
+}
+.dropdown-trigger .icon {
+  flex: 0 0 auto;
+  width: 1rem;
+  height: 1rem;
+  margin-bottom: calculateRem(2px);
+  margin-left: $inuit-base-spacing-unit--tiny;
+}
+.dropdown-content {
```
Comment (mdwragg) [inline sass/px-dropdown.scss]: Moar BEM.

File: sass/px-dropdown.scss [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  background-color: var(--px-dropdown-bg-color, $white);
+  border: 1px solid $gray3;
+  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.15);
+  max-height: var(--px-dropdown-max-height, 300px);
+  max-width: var(--px-dropdown-max-width, 400px);
+  word-break: break-word;
+}
+.dropdown-option {
```
Comment (mdwragg) [inline sass/px-dropdown.scss]: And here too (BEM)

File: px-dropdown.html [hunk 4]
```diff
@@ -129,478 +127,537 @@
...
-      if (!this.preventCloseOnOutsideClick && !this._selfInPath(Polymer.dom(evt).path) &&
-        this.opened) {
-
-        var content = Polymer.dom(this).querySelector("px-dropdown-content");
-        this._flipOpened();
-        content.close();
-        this._reset();
-      }
+    _getElement: function(target) {
+      return document.querySelector(target);
```
Comment (mdwragg) [inline px-dropdown.html]: Should drive this through the polymer API to make sure it work in shady/shadow dom.

File: px-dropdown.html [hunk 5]
```diff
@@ -129,478 +127,537 @@
...
-      //find the element if we have been passed an id
-      if (typeof boundTarget === 'string') {
-        this.boundTarget = this.domHost ? this.domHost.$[boundTarget] :
-          Polymer.dom(this.ownerDocument).querySelector('#' + boundTarget);
+    _itemsChanged: function(items) {
+      if(items[0] && typeof items[0] === 'string') {
+        var length = items.length,
+            computedItems = [],
+            i;
+        for(i=0; i<length; i++) {
```
Comment (mdwragg) [inline px-dropdown.html]: Feel free to use the more semantically safe/easy to understand .forEach et al interators here, I doubt we're talking about large numbers of items here to perf is probably not a huge worry.

File: px-dropdown.html [hunk 6]
```diff
@@ -129,478 +127,537 @@
...
-        dropcell = this.contentAnchor ? this.contentAnchor : this.$.dropcell,
-        dropcellRect = dropcell.getClientRects()[0],
-        newTop = (parseInt(dropdown.offsetTop) - parseInt(dropdownRect.height) - parseInt(dropcellRect.height)) + 'px';
-
-      dropdown.style.top = newTop;
-      this.above = true;
+    _setFocusedOption: function(newOption, oldOption) {
+      this._focusedOption = newOption;
+      if(newOption) {
+        newOption.classList.add('focused');
```
Comment (mdwragg) [inline px-dropdown.html]: Prefer the use of `this.toggleClass()` rather than manually manipulating the classList .

File: px-dropdown.html [hunk 7]
```diff
@@ -129,478 +127,537 @@
...
-              var content = Polymer.dom(this).querySelector('px-dropdown-content');
-              content._handleKeyEvent('ENTER', event);
-            } else {
-              this.triggerClicked(event);
-              event.detail.keyboardEvent.preventDefault();
-            }
-            break;
-        }
+      this._keyboardBeingUsed = true;
+      var keyPressed = event.detail.combo,
```
Comment (mdwragg) [inline px-dropdown.html]: Make sure events are normalised to work in shadow/shady DOM.

File: px-dropdown.html [hunk 8]
```diff
@@ -129,478 +127,537 @@
...
+          break;
+        case 'up':
+          // If closed, open it
+          if(!this.opened) {
+            this.toggle();
+          }
+          // If something is focused, move to the previous sibling
+          if(focused > 0) {
+            this._setFocusedOption(options[focused - 1], options[focused]);
+            this.$.dropdown.querySelector('.dropdown-content').scrollTop -= options[focused].offsetHeight;
```
Comment (mdwragg) [inline px-dropdown.html]: Drive through Polymer dom API to make sure safe across shadow/shady dom.

File: px-dropdown.html [hunk 9]
```diff
@@ -129,478 +127,537 @@
...
-     *   key: "key of the selected element" }
-     * ```
-     * @event px-dropdown-value-changed
+     * Event handler when the mouse hovers over a dropdown list item.
      */
+    _hoverOn: function (event) {
+      if (!this._keyboardBeingUsed) {
+        var currHighlightedItem = this.querySelector('.dropdown-option.focused');
+        if (currHighlightedItem) {
+          currHighlightedItem.classList.remove('focused');
```
Comment (mdwragg) [inline px-dropdown.html]: As above, prefer `this.toggleClass`.

File: sass/px-dropdown.scss [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+.dropdown-trigger {
+  outline: none;
+  max-width: inherit;
+  width: inherit;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  justify-content: space-between;
+}
+.dropdown-trigger .label {
```
Comment (mdwragg) [inline sass/px-dropdown.scss]: As above, more BEM loveliness.

File: sass/px-dropdown.scss [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  overflow: hidden;
+  text-overflow: ellipsis;
+  text-align: left;
+}
+.dropdown-trigger .icon {
```
Comment (mdwragg) [inline sass/px-dropdown.scss]: As above, more BEM loveliness.

**Comment by randyaskin** [comment]
To summarize our Slack conversation:

- BEM isn't possible, since iron-dropdown is expecting `.dropdown-trigger` and `.dropdown-content`
- iron-a11y-keys seems to take care of normalization with the `target` property, but I've updated the others
- replaced `for` loop with `forEach` and `classList` manipulation with `toggleClass` as requested

**Pull request merged**